File: taller1/factorial.c
#include <stdio.h>
#include <time.h>

double factorial_recursivo(int n);

double factorial_recursivo(int n) {
    /*
    Usa 'double' para poder manejar resultados más grandes, aunque con pérdida de precisión.
    */
    if (n == 0) {
        return 1.0;
    } 
    else {
        return (double)n * factorial_recursivo(n - 1);
    }
}

int main() {
    int numero = 10000; 

    // --- Medición de tiempo ---
    clock_t inicio, fin;
    double tiempo_cpu_usado;

    inicio = clock(); // Guarda el tiempo de inicio

    double resultado = factorial_recursivo(numero);
    
    fin = clock(); // Guarda el tiempo de finalización

    // Calcula el tiempo transcurrido en segundos
    tiempo_cpu_usado = ((double) (fin - inicio)) / CLOCKS_PER_SEC;

    printf("Cálculo completado.\n");
    //printf("El factorial de %d es aproximadamente: %e\n", numero, resultado); // Notación científica
    printf("El programa en C tardó: %f segundos.\n", tiempo_cpu_usado);

    return 0;
}

File: taller1/factorial.py
import time
import sys

# Aumenta el límite de recursión de Python para números grandes.
sys.setrecursionlimit(200000)

def factorial_recursivo(n):
    if n == 0:
        return 1
    else:
        # print(f"Calculando para {n}...") 
        return n * factorial_recursivo(n - 1)

if __name__ == "__main__":
    numero = 100000 

    # --- Medición de tiempo ---
    inicio = time.perf_counter() # Obtiene un tiempo de alta precisión

    try:
        resultado = factorial_recursivo(numero)
        fin = time.perf_counter() # Obtiene el tiempo al finalizar
        
        # print(f"El factorial de {numero} es: {resultado}") # PD: Esto explota con numeros grandes xd
        print(f"Cálculo completado.")
        print(f"El programa en Python tardó: {fin - inicio:0.6f} segundos.")

    except RecursionError:
        print(f"Error: Se ha excedido el límite de recursión para el número {numero}.")
    except Exception as e:
        print(f"Ocurrió un error: {e}")